libraries/Microsoft.Bot.Builder.Azure/AzureTableStorageStateStorageProvider.cs

```diff
@@ -0,0 +1,271 @@
+        private const string BotRawKeyTableStoragePropertyName = BotPropertyPrefix + "RawKey";
+
+        private static readonly HashSet<string> CheckedTables = new HashSet<string>();
+        private static readonly Regex TableStorageKeyInvalidCharactersRegex = new Regex("[!@#$%^&*\\(\\)~/\\\\><,.?';\"\u0000-\u001F\u007F-\u009F]", RegexOptions.Singleline | RegexOptions.CultureInvariant | RegexOptions.Compiled);
+
+        private readonly CloudTable _table;
+
+        public AzureTableStorageStateStorageProvider(string dataConnectionString, string tableName) : this(CloudStorageAccount.Parse(dataConnectionString), tableName)
+        {
+        }
+
+        public AzureTableStorageStateStorageProvider(CloudStorageAccount storageAccount, string tableName)
+        {
+            var tableClient = storageAccount.CreateCloudTableClient();
+            _table = tableClient.GetTableReference(tableName);
+
+            if (CheckedTables.Add($"{storageAccount.TableStorageUri.PrimaryUri.Host}-{tableName}"))
+            {
+                _table.CreateIfNotExistsAsync().Wait();
+            }
+        }
+
+        public CloudTable Table => _table;
+
+        public IStateStorageEntry CreateNewEntry(string stateNamespace, string key)
+        {
+            var tableEntity = new DynamicTableEntity(NormalizeKeyValueForTableStorage(stateNamespace), NormalizeKeyValueForTableStorage(key));
+            tableEntity.Properties[BotRawStateNamespaceTableStoragePropertyName] = new EntityProperty(stateNamespace);
+            tableEntity.Properties[BotRawKeyTableStoragePropertyName] = new EntityProperty(key);
+
+            return new AzureTableStorageEntityStateStorageEntry(tableEntity);
+        }
+
+        public async Task<IEnumerable<IStateStorageEntry>> Load(string stateNamespace)
+        {
+            var queryContinuationToken = default(TableContinuationToken);
+            var query = new TableQuery()
+                            .Where(TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.Equal, NormalizeKeyValueForTableStorage(stateNamespace)));
+
+
+            var results = new List<IStateStorageEntry>();
+
+            do
+            {
+                var entitySegment = await _table.ExecuteQuerySegmentedAsync(query, queryContinuationToken).ConfigureAwait(false);
+
+                results.AddRange(entitySegment.Results.Select(entity => new AzureTableStorageEntityStateStorageEntry(entity)));
+
+                queryContinuationToken = entitySegment.ContinuationToken;
+            } while (queryContinuationToken != default(TableContinuationToken));
+
+            return results;
+        }
+
+        public async Task<IStateStorageEntry> Load(string stateNamespace, string key)
+        {
+            var tableResult = await _table.ExecuteAsync(TableOperation.Retrieve(NormalizeKeyValueForTableStorage(stateNamespace), NormalizeKeyValueForTableStorage(key))).ConfigureAwait(false);
```

Comment:
>NormalizeKeyValueForTableStorage(stateNamespace) [](start = 80, length = 48)

This inherently means that all Conversation state is in 1 partition which severely limits the scalability of table storage provider.  That may be OK, as maybe we should be pushing people to using blob or cosmosdb instead.  Is the namespace hierarchial? (I know that you were pushing for that in one of our meetings)  That would allow you to have blob prefix queries for blob storage or to leverage in cosmosdb for queries...